File: beecrowd/1527/uf.py
ADD_OP = 1
BATTLE_OP = 2
RAFAEL_SET = 1

def find(sets, s):
    rep = s
    while sets[rep] != rep:
        rep = sets[rep]
    while sets[s] != rep:
        sets[s], s = rep, sets[s]
    return rep

def union(sets, s1, s2, sizes, points):
    r1 = find(sets, s1)
    r2 = find(sets, s2)
    if sizes[r2] > sizes[r1]:
        r1, r2 = r2, r1
    sets[r2] = r1
    sizes[r1] += sizes[r2]
    points[r1] += points[r2]
    return r1

def battle(sets, s1, s2, points):
    r1 = find(sets, s1)
    r2 = find(sets, s2)
    winner = r1 if points[r1] > points[r2] else r2
    return int(winner == find(sets, RAFAEL_SET))

def main():
    while True:
        n, m = map(int, input().split())
        if n == m == 0:
            return

        points = [0] + [int(x) for x in input().split()]
        sets = [i for i in range(n+1)]
        sizes = [1] * (n + 1)
        rafael_wins = 0

        for _ in range(m):
            op, a, b = map(int, input().split())
            if op == ADD_OP:
                union(sets, a, b, sizes, points)
            elif op == BATTLE_OP:
                rafael_wins += battle(sets, a, b, points)
        print(rafael_wins)

if __name__ == '__main__':
    main()

File: beecrowd/1527/uf.cpp
#include <bits/stdc++.h>
using namespace std;
using vi = vector<int>;

const int ADD_OP = 1, BATTLE_OP = 2, RAFAEL_SET = 1;

int find(vi& sets, int s) {
  int tmp, rep = s;
  while (sets[rep] != rep)
    rep = sets[rep];
  while (sets[s] != s) {
    tmp = sets[s];
    sets[s] = rep;
    s = tmp;
  }
  return rep;
}

int union_sets(vi& sets, vi& sizes, vi& points, int s1, int s2) {
  int r1 = find(sets, s1), r2 = find(sets, s2);
  if (sizes[r2] > sizes[r1])
    swap(r1, r2);
  sets[r2] = r1;
  sizes[r1] += sizes[r2];
  points[r1] += points[r2];
  return r1;
}


int battle(vi& sets, vi& points, int s1, int s2) {
  int r1 = find(sets, s1), r2 = find(sets, s2);
  if (points[r1] == points[r2]) return 0;
  int winner = points[r1] > points[r2] ? r1 : r2;
  return winner == find(sets, RAFAEL_SET);
}

int main() {
  int n, m, rafael_wins, op, a, b, i;
  while (true) {
    cin >> n >> m;
    if (n == 0 && m == 0) break;

    vi sets(n+1), points(n+1), sizes(n+1, 1);
    rafael_wins = 0;
    for (i = 1; i <= n; i++) {
      cin >> points[i];
      sets[i] = i;
    }

    for (i = 0; i < m; i++) {
      cin >> op >> a >> b;
      if (op == ADD_OP)
        union_sets(sets, sizes, points, a, b);
      else if (op == BATTLE_OP)
        rafael_wins += battle(sets, points, a, b);
    }
    cout << rafael_wins << '\n';
  }
}
